### PR TITLE
Unresolved dependency warning includes source position. Fixes #528

### LINKED ISSUE #528: Error message for missing plugin is suboptimal
I just tried to load some project with sbt 0.12.0. Sbt failed to resolve some plugin dependency and produced rather unhelpful message (attached below).

Let me make clear what's suboptimal with this error message:
- it provides a long stacktrace that is most likely not helpful to the user
- it doesn't provide any clue where this plugin dependency has been defined (in this case it was in `~/.sbt/plugins/plugins.sbt`

The way the error is presented makes it really hard for the user to figure out a solution because it's fairly easy to forget about global plugin configuration. The best solution would be to get rid of the stacktrace and print the path of file where given dependency has been defined.

```

[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.typesafe.sbteclipse#sbteclipse-plugin;2.1.0-RC1: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.
[warn]      com.typesafe.sbteclipse:sbteclipse-plugin:2.1.0-RC1 (sbtVersion=0.12, scalaVersion=2.9.2)
[warn] 
sbt.ResolveException: unresolved dependency: com.typesafe.sbteclipse#sbteclipse-plugin;2.1.0-RC1: not found
    at sbt.IvyActions$.sbt$IvyActions$$resolve(IvyActions.scala:211)
    at sbt.IvyActions$$anonfun$update$1.apply(IvyActions.scala:122)
    at sbt.IvyActions$$anonfun$update$1.apply(IvyActions.scala:121)
    at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:114)
    at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:114)
    at sbt.IvySbt$$anonfun$withIvy$1.apply(Ivy.scala:102)
    at sbt.IvySbt.liftedTree1$1(Ivy.scala:49)
    at sbt.IvySbt.action$1(Ivy.scala:49)
    at sbt.IvySbt$$anon$3.call(Ivy.scala:58)
    at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:75)
    at xsbt.boot.Locks$GlobalLock.withChannelRetries$1(Locks.scala:58)
    at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:79)
    at xsbt.boot.Using$.withResource(Using.scala:11)
    at xsbt.boot.Using$.apply(Using.scala:10)
    at xsbt.boot.Locks$GlobalLock.liftedTree1$1(Locks.scala:51)
    at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:51)
    at xsbt.boot.Locks$.apply0(Locks.scala:30)
    at xsbt.boot.Locks$.apply(Locks.scala:27)
    at sbt.IvySbt.withDefaultLogger(Ivy.scala:58)
    at sbt.IvySbt.withIvy(Ivy.scala:99)
    at sbt.IvySbt.withIvy(Ivy.scala:95)
    at sbt.IvySbt$Module.withModule(Ivy.scala:114)
    at sbt.IvyActions$.update(IvyActions.scala:121)
    at sbt.Classpaths$$anonfun$work$1$1.apply(Defaults.scala:949)
    at sbt.Classpaths$$anonfun$work$1$1.apply(Defaults.scala:947)
    at sbt.Classpaths$$anonfun$doWork$1$1$$anonfun$54.apply(Defaults.scala:970)
    at sbt.Classpaths$$anonfun$doWork$1$1$$anonfun$54.apply(Defaults.scala:968)
    at sbt.Tracked$$anonfun$lastOutput$1.apply(Tracked.scala:35)
    at sbt.Classpaths$$anonfun$doWork$1$1.apply(Defaults.scala:972)
    at sbt.Classpaths$$anonfun$doWork$1$1.apply(Defaults.scala:967)
    at sbt.Tracked$$anonfun$inputChanged$1.apply(Tracked.scala:45)
    at sbt.Classpaths$.cachedUpdate(Defaults.scala:975)
    at sbt.Classpaths$$anonfun$45.apply(Defaults.scala:855)
    at sbt.Classpaths$$anonfun$45.apply(Defaults.scala:852)
    at sbt.Scoped$$anonfun$hf10$1.apply(Structure.scala:586)
    at sbt.Scoped$$anonfun$hf10$1.apply(Structure.scala:586)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:49)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$12.apply(Structure.scala:311)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$12.apply(Structure.scala:311)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$5.work(System.scala:71)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:232)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:232)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:18)
    at sbt.Execute.work(Execute.scala:238)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:232)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:232)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:30)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
    at java.lang.Thread.run(Thread.java:680)
[error] (*:update) sbt.ResolveException: unresolved dependency: com.typesafe.sbteclipse#sbteclipse-plugin;2.1.0-RC1: not found
```

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #528.

Unresolved dependency warning is moved to `UnresolvedDependencyWarning` class including
the fail path that was added in #1467.
To display the source position, I need to access the State, so I had to move the
error processing out of `IvyActions` and add `UnresolvedDependencyWarning`, which is
aware of State.

Here's global plugin dependency sample:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: com.typesafe.sbt#sbt-pgp;0.100.1: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.
[warn]      com.typesafe.sbt:sbt-pgp:0.100.1 (sbtVersion=0.13, scalaVersion=2.10)
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      com.typesafe.sbt:sbt-pgp:0.100.1 (sbtVersion=0.13, scalaVersion=2.10) (/Users/foo/dotfiles/sbt/0.13/plugins/gpg.sbt#L3-4)
[warn]        +- org.scala-sbt:global-plugins:0.0 (sbtVersion=0.13, scalaVersion=2.10)
```

Normal library dependency sample:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: foundrylogic.vpp#vpp;2.2.1: not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]      foundrylogic.vpp:vpp:2.2.1
[warn]        +- org.apache.cayenne:cayenne-tools:3.0.2
[warn]        +- org.apache.cayenne.plugins:maven-cayenne-plugin:3.0.2 (/Users/foo/hello/build.sbt#L28)
[warn]        +- d:d_2.10:0.1-SNAPSHOT
```

/review @jsuereth /cc @gkossakowski

### CODE REVIEW

File: main/src/main/scala/sbt/UnresolvedDependencyWarning.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    // This is a mapping between modules and original position, in which the module was introduced.
+    lazy val modulePositions: Seq[(ModuleID, SourcePosition)] =
+      try {
+        stateOpt map { state =>
+          val extracted = (Project extract state)
+          val sk = (Keys.libraryDependencies in (GlobalScope in extracted.currentRef)).scopedKey
+          val empty = extracted.structure.data set (sk.scope, sk.key, Nil)
+          val settings = extracted.structure.settings filter { s: Setting[_] =>
+            (s.key.key == Keys.libraryDependencies.key) &&
+              (s.key.scope.project == Select(extracted.currentRef))
```
Comment (jsuereth) [inline main/src/main/scala/sbt/UnresolvedDependencyWarning.scala]: Current ref may not be the reference to the project in which update is running.  Have you tried this on a multi-project build?

File: main/src/main/scala/sbt/UnresolvedDependencyWarning.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  val withExtra: Seq[ModuleID],
+  val failedPaths: Seq[Seq[(ModuleID, Option[String])]])
+
+private[sbt] object UnresolvedDependencyWarning {
+  // This method used to live under IvyActions.scala, but it's moved here because it's now aware of State.
+  private[sbt] def apply(err: ResolveException, stateOpt: Option[State]): UnresolvedDependencyWarning = {
+    // This is a mapping between modules and original position, in which the module was introduced.
+    lazy val modulePositions: Seq[(ModuleID, SourcePosition)] =
+      try {
+        stateOpt map { state =>
```
Comment (jsuereth) [inline main/src/main/scala/sbt/UnresolvedDependencyWarning.scala]: Why don't you pass all this in via some "config" option rather than via state?

**Comment by jsuereth** [comment]
So, this looks clever and nice for the user.  Major concern is your use of projectRef.   I believe there's a way, inside a setting, to grab the scope of the setting itself (see how the sourceDirectory settings are generated).   The way you're currently using state is a bit suspect though, but generally I think it's the only way to grab the information you want.

What may be nicer, architecturally, is some kind of state post-process step which auto-generates a "LibraryDependencySourceIndex" setting you can just depend on.   This setting could hide the complexities of looking up source positions.

Ideally, we don't want the cost of scanning through settings to be hit EVERY TIME you run resolution, so we may even want to try to cache such an index after load time and avoid altering it until the libraryDependencies settings change.

**Comment by eed3si9n** [comment]
> Major concern is your use of projectRef.

You are right about not trusting `extracted.currentRef`. I'm now using `thisProjectRef` instead, and that one seems to work.

> What may be nicer, architecturally, is some kind of state post-process step which auto-generates a "LibraryDependencySourceIndex" setting you can just depend on.

Agreed. With the new design, that `Module -> SourcePosition` index is called `UnresolvedWarningConfiguration`, which is calculated using a new task `unresolvedWarningConfiguration`. 

> Ideally, we don't want the cost of scanning through settings to be hit EVERY TIME you run resolution, so we may even want to try to cache such an index after load time and avoid altering it until the libraryDependencies settings change.

`unresolvedWarningConfigurationTask` does just that:

``` scala
    val f = Tracked.inputChanged(cacheFile / "input_uwc") { (inChanged: Boolean, in: Seq[ModuleID]) =>
      val outCache = Tracked.lastOutput[Seq[ModuleID], UnresolvedWarningConfiguration](outCacheFile) {
        case (_, Some(out)) if !inChanged => out
        case _                            => UnresolvedWarningConfiguration(modulePositions)
      }
      outCache(in)
    }
    f(libraryDependencies.value)
```

**Comment by jsuereth** [comment]
So now may only complaint is that `UnresolvedWarningConfiguration` is a name about what you're trying to do but not what the configuration actually IS.  It's not actually the configuration for unresolved warnings, but a "LibraryDependencyDefinitionIndex" or some such.  Other than petty naming discussions, LGTM.

**Comment by eed3si9n** [comment]
Added `dependencyPositions` task to cache dependency positions explicitly.

**Comment by jsuereth** [comment]
Cool.  LGTM

**Pull request merged**